### PR TITLE
[FLINK-20191][document] Add documents for FLIP-95 ability interfaces

### PR DESCRIPTION
**Pull request by fsk119**
## What is the purpose of the change

Add document for the abilities interface.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

File: docs/dev/table/sourceSinks.zh.md [hunk 1]
```diff
@@ -193,6 +192,149 @@ for more information.
 The runtime implementation of a `LookupTableSource` is a `TableFunction` or `AsyncTableFunction`. The function
 will be called with values for the given lookup keys during runtime.
 
+#### Defining a Dynamic Table Source with Projection Push-Down
```
Comment (twalthr) [inline docs/dev/table/sourceSinks.zh.md:195]: How about we just describe the interfaces in 1 or 2 sentences. Maybe in a table. And link to the corresponding class on Github. It is very difficult to keep docs and interfaces in sync. We should avoid code duplication.

**Comment by fsk119** [comment]
I have updated the doc. Currently it uses table to record abilities interfaces.
image

File: docs/dev/table/connectors/kafka.md [hunk 2]
```diff
@@ -249,6 +249,15 @@ Besides enabling Flink's checkpointing, you can also choose three different mode
 
 Please refer to [Kafka documentation]({% link dev/connectors/kafka.md %}#kafka-producers-and-fault-tolerance) for more caveats about delivery guarantees.
 
+### Per-partition-watermark Source
+
+Flink supports to emit per-partition-watermark for Kafka. Using this feature, watermarks are generated inside the Kafka consumer. The per-partition-watermark are merged in
+the same way as watermarks are merged on stream shuffles. The output watermark of the source is determined by the minimum watermark among the partitions it reads. Considering a watermark assigner
+advance the watermark according to the event-time on the records. If some partitions in the topics are idle, the watermark assigner will not advance. You can alleviate this problem by
+setting appropriate [idelness timeouts]({{ site.baseurl }}/dev/event_timestamps_watermarks.html#dealing-with-idle-sources).
```
Comment (twalthr) [inline docs/dev/table/connectors/kafka.md:257]: This documentation makes no sense for Table/SQL users. We don't have a possibility to set this via options, right?

Comment (twalthr) [reply docs/dev/table/connectors/kafka.md:257]: We should actually introduce an option for it.

File: docs/dev/table/sourceSinks.zh.md [hunk 3]
```diff
@@ -31,16 +31,15 @@ of a dynamic table is stored in external systems (such as databases, key-value s
 _Dynamic sources_ and _dynamic sinks_ can be used to read and write data from and to an external system. In
 the documentation, sources and sinks are often summarized under the term _connector_.
 
-Flink provides pre-defined connectors for Kafka, Hive, and different file systems. See the [connector section]({% link dev/table/connectors/index.zh.md %})
+Flink provides pre-defined connectors for Kafka, Hive, and different file systems. See the [connector section]({% link dev/table/connectors/index.md %})
```
Comment (twalthr) [inline docs/dev/table/sourceSinks.zh.md:34]: undo this change?

File: docs/dev/table/sourceSinks.zh.md [hunk 4]
```diff
@@ -31,16 +31,15 @@ of a dynamic table is stored in external systems (such as databases, key-value s
...
 for more information about built-in table sources and sinks.
 
 This page focuses on how to develop a custom, user-defined connector.
 
 <span class="label label-danger">Attention</span> New table source and table sink interfaces have been
 introduced in Flink 1.11 as part of [FLIP-95](https://cwiki.apache.org/confluence/display/FLINK/FLIP-95%3A+New+TableSource+and+TableSink+interfaces).
-Also the factory interfaces have been reworked. FLIP-95 is not fully implemented yet. Many ability interfaces
-are not supported yet (e.g. for filter or partition push down). If necessary, please also have a look
-at the [old table sources and sinks page]({% link dev/table/legacySourceSinks.zh.md %}). Those interfaces
+Also the factory interfaces have been reworked. Many ability interfaces has been supported (e.g. for filter or partition push down).
```
Comment (twalthr) [inline docs/dev/table/sourceSinks.zh.md:41]: Remove `Many ability interfaces has been supported (e.g. for filter or partition push down).` entirely?

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the PR @fsk119. I gave some feedback.

File: docs/dev/table/connectors/kafka.md (hunk 2 above)
Comment (fsk119) [reply docs/dev/table/connectors/kafka.md:257]: Could you give a bad case about this feature? With this feature, we can get more fine-grained watermark. Only some partition are idle, the watermark assigner doesn't advance itself. However, we can set idle timeout to solve this problem.

Comment (twalthr) [reply docs/dev/table/connectors/kafka.md]: Maybe there is some misunderstanding. The link that you put in the docs helps DataStream API users:
```
WatermarkStrategy
        .<Tuple2<Long, String>>forBoundedOutOfOrderness(Duration.ofSeconds(20))
        .withIdleness(Duration.ofMinutes(1));
```

But it should be updated to help Table/SQL API users.

I found `org.apache.flink.table.api.config.ExecutionConfigOptions#TABLE_EXEC_SOURCE_IDLE_TIMEOUT` we should link to this instead if it works for the Kafka connector.

Comment (fsk119) [reply docs/dev/table/connectors/kafka.md]: You are right. I modify the link: [table.exec.source.idle-timeout](https://ci.apache.org/projects/flink/flink-docs-master/dev/table/config.html#table-exec-source-idle-timeout) and [Kafka watermark strategies](https://ci.apache.org/projects/flink/flink-docs-master/dev/event_timestamps_watermarks.html#watermark-strategies-and-the-kafka-connector).

**Comment by twalthr** [comment]
Thanks for the update. I will take care of merging this...

**Review by twalthr — APPROVED** [review]
Thanks for the update. I will take care of merging this...

**Comment by fsk119** [comment]
Thanks for your feedback. Cheers ^_^
